### PR TITLE
Add eye_like

### LINKED ISSUE #9244: [pytorch] [feature request] torch.eye_like
Would be useful for regularisation code.

Currently, to create an identity matrix on the same device we need to pass both `device` and `dtype` explicitly to `torch.eye`.

Maybe all dimensions except the last two, can be considered as batch dimensions.

### PR DESCRIPTION
**Pull request by vishwakftw**
Closes #9244 

Added doc string and tests.

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -2248,6 +2248,13 @@ def test_eye(self):
         torch.eye(100, 100, out=res2)
         self.assertEqual(res1, res2)
 
+    def test_eye_like(self):
+        res1 = torch.eye(100, 100)
```
Comment (zou3519) [inline test/test_torch.py]: Could you test the following cases please:
- torch.eye_like(long-rectangle-matrix)
- torch.eye_like(fat-rectangle-matrix)

**Review by ezyang — APPROVED** [review]

File: test/test_torch.py (hunk 1 above)
Comment (ezyang) [reply test/test_torch.py]: Isn't this backending to `eye`?

**Comment by vadimkantorov** [comment]
Currently this PR supports only 2-d inputs. Are there thoughts about supporting n-d by treating any leading dimensions as batch dimensions (can either unsqueeze+expand over additional dimensions or materialize the full tensor)?

File: aten/src/ATen/native/TensorFactories.cpp [hunk 2]
```diff
@@ -160,20 +160,28 @@ Tensor& eye_out_cpu(Tensor& result, int64_t n, int64_t m) {
   return result;
 }
 
+Tensor eye_like(const Tensor& self) {
+  AT_CHECK(self.dim() == 2, "Expected input to be two-dimensional, instead got ",
+           self.dim(), " dimensions");
+  return native::eye(self.size(0), self.size(1), self.options());
```
Comment (zou3519) [inline aten/src/ATen/native/TensorFactories.cpp]: Does this work for sparse? (What is the error message, if there is one?)

**Review by zou3519 — COMMENTED** [review]
looks good! I had some minor comments & questions

File: test/test_torch.py (hunk 1 above)
Comment (zou3519) [reply test/test_torch.py]: True

**Comment by gchanan** [comment]
I'd rather not include this function as a built-in because it's different enough from the other _like functions (which work across all shapes) to be misleading.  If we were to support N-dimensional eye(s) then it would be okay, but I don't think we have any rank 2+ functions defined for which identity makes sense.

**Comment by vadimkantorov** [comment]
@gchanan My usecase is a batch of covariance matrices to which I'd like to add epsilon times diagonal for regularization. Then when batched `potrf`, `trtrs` and others are implemented (https://github.com/pytorch/pytorch/issues/7500), this would become nice for some simple multivariate normal likelihood computations

**Comment by gchanan** [comment]
@vadimkantorov just to clarify, you are talking about the your previous comment about accepting n-dimensional inputs and returning batched identity outputs?  If so, that seems useful, but I don't know why we'd call that `eye_like` over something like `batched_identity`, `batched_eye_like` or something.

**Comment by vadimkantorov** [comment]
@gchanan: accepting `B x D x D` and returning `B x D x D` as well.

I don't have an opinion about naming (here explicit batch-prefix were considered not primary choice  https://github.com/pytorch/pytorch/pull/9102).

File: aten/src/ATen/native/TensorFactories.cpp (hunk 2 above)
Comment (vishwakftw) [reply aten/src/ATen/native/TensorFactories.cpp]: It doesn't work for sparse. The error message falls back to the one printed by `eye_out`, but I have added it here too.

**Comment by vishwakftw** [comment]
Is this good to go? Sorry about the reminder.

**Comment by SsnL** [comment]
@gchanan any more comments?

**Comment by soumith** [comment]
I really dont like the `eye_like` name to imply something magical in terms of expected semantics. I share Greg's concerns.

Unfortunately, the PR is implemented (with docs/tests), so its a bit painful to reject it, but I think that's the right call.

**Comment by vishwakftw** [comment]
I share @gchanan 's concerns too. In fact in the issue, I did give a one line solution for `eye_like` which should have worked just as well. This PR wasn't all that hard to implement, so if it comes up later on, I can re-trace the steps. Closing it due to the ambiguities, and sorry for any inconvenience caused.

**Comment by vadimkantorov** [comment]
@soumith I think it's worth to see actual usages of `torch.eye` in `torch.distributions` and `Pyro`.

If many usages are meaning to do sth like batched `torch.eye_like`, it may be a reason to reconsider.

Basically, my grievance is that there's no concise way to transfer type and device (and potentially layout) to tensor factory methods (there seems no concice equivalent of TensorOptions in Python): `torch.eye(some_shape, device = x.device, type = x.dtype)` without calling `type_as` and resorting to copying or a less readable `torch.eye(some_shape, out = x.new())`

**Comment by ezyang** [comment]
@vadimkantorov Your grievance is legit, and I remember mentioning this to Greg a while back.

Basically, we want this: `torch.eye(some_shape, **x.options())` (or some similar name; options is what we're calling it in C++). This would be a very useful addition, and should not be hard to code at al.

**Comment by SsnL** [comment]
While we are on this topic, how about `new_eye` instead of `eye_like`?  Since `new_eye` always takes in shape (or size of the diagonal in this case), it won't be ambiguous.

Related: should we have `new_randn` and `new_rand`? I find them rather helpful in writing device-agnostic model.

**Comment by ezyang** [comment]
@SsnL Wasn't our actual plan to make, e.g., randn accept device, so no need for a new function?

**Comment by SsnL** [comment]
@ezyang it already takes in device, right? I thought the discussion was
about avoiding the verbose `device=t.device, dtype=t.dtype` kwarg.

**Comment by ezyang** [comment]
Right. So as long as you can get the dict `{device: t.device, dtype: t.dtype}` conveniently, you can just `**` your way to victory

**Comment by SsnL** [comment]
@ezyang yeah, but `**t.options()` is still quite verbose, and not immediately readable to people not familiar with pytorch. In addition, this pretty much means that `Tensor.options()` will pretty much have only this use. Moreover, the purpose of the `new_` functions is to preserve device and dtype (and those are overwritable too!). So, with `new_` functions so suitable, having a separate mechanism seems unnecessary.

**Comment by vadimkantorov** [comment]
@SsnL `tensor.new_ones(...)` is less verbose than `torch.ones_like(..., **tensor.options())`, but is not very intuitive/readable to new users IMO either. Another variant: `torch.ones(..., type_as = tensor)` or `torch.ones(..., as = tensor)`?

**Comment by apaszke** [comment]
`torch.ones(..., like=tensor)`?

**Comment by gchanan** [comment]
I like the `**options()` because it matches the C++ code.

**Comment by apaszke** [comment]
@gchanan but it's really verbose, and I don't think we should use limitations of C++ as our golden standard when designing APIs.

**Comment by gchanan** [comment]
I agree it's verbose, but it's discoverable (because it's the same function) and I understand what it's doing the first time I see it.

With the `new_XXX` functions it's not clear what `new` refers to -- those are just named that way because of the deprecated `new` function -- someone coming to PyTorch won't understand what `new` means.

Similarly with,`like=`...like usually refers to shape / dtype / layout / device -- but I presume here it won't include the shape (if I'm using it with e.g. `torch.ones`).  That's subtle (either if you include it or not), which means I have to read the documentation, internalize it, etc.

**Comment by vadimkantorov** [comment]
`torch.ones(..., like/as/type_as/... = tensor)` and `torch.ones(..., **tensor.options())` need not be exclusive

The `like/as` variant is really concies; and the `**tensor.options()` suppport just boils down to supporting `options()` method, the keyword arguments like device/dtype are already there.

**Comment by gchanan** [comment]
This seems to boil down to whether you prefer being concise or having the semantics be clearer/more consistent.  I prefer the later.

**Comment by SsnL** [comment]
I agree with @apaszke on this. We shouldn’t limit our much widely used
Python API by the constraint of c++, especially when the c++ API is much
less used, still experimental (i.e. c++ API can be changed much more easily
w/o considering BC), and already different with the Python API in many
ways.

**Comment by SsnL** [comment]
Moreover, I don’t think tensor.options() is clear either. For example, why
are strides, sizes, offsets, whether in pinned memory, requires_grad, etc.
not considered as “options()”?

**Comment by gchanan** [comment]
> I agree with @apaszke on this. We shouldn’t limit our much widely used
Python API by the constraint of c++, especially when the c++ API is much
less used, still experimental (i.e. c++ API can be changed much more easily
w/o considering BC), and already different with the Python API in many
ways.

I don't really see how the limitations of C++ forced us to adopt the `options` API except so far as it's not a kwarg argument.  We could implement `like` as an argument just as easily in C++ (again, with it lacking being a kwarg).

> Moreover, I don’t think tensor.options() is clear either. For example, why
are strides, sizes, offsets, whether in pinned memory, requires_grad, etc.
not considered as “options()”?

Because those are never included in the factory APIs (modulo requires_grad, and I'm not sure what the story is there, but IMO we shouldn't have requires_grad in any of the factory APIs anyway).  I.e. I'm imaging documentation such as:
```
ones(shape, *,  dtype, device, layout)
ones(shape, **options)
```
It's seems pretty clear options encapsulates the kwarg-only arguments.

You could make the same argument for having a `like` kwarg-only arg except `like` usually incoldes the shape, so now you need to explain that `like` means different things in different contexts.

**Comment by SsnL** [comment]
What I tried to say is that
1. `**t.options()` isn't more consistent. 
    If we really want to be consistent, the Python API would be like `torch.randn((1, 2, 3), options=torch.TensorOptions(torch.cuda).dtype(torch.double))`. We should value Python API's conciseness and clarity over consistency with experimental C++ API any day.
2. `**t.options()` isn't clearer either. 
   As you said, we would need to write the documentation clearly if we do this. However, didn't we decide against `new_*` because we want users to directly understand the code without being familiar with pytorch? If we want to assume that users have read the doc, why don't we just stick with the `new_*` variants, whose whole purpose is to preserve `dtype`, `device` and `layout`?
3. Again, what's the use of `t.options()` as a public API if its use is pretty much only in kwarg expansion of tensor factory methods? 
4.
   ```
ones(shape, *,  dtype, device, layout)
ones(shape, **options)
```
doesn't support overwriting `dtype`, `device` or `layout`.

I'm not saying that additional kwargs like `as` or `like` is better. IMHO, they all require users to understand somewhat about Pytorch as they aren't immediately clear what is preserved and what is not. Given this, I still believe that `new_*` might just be better than these two because `new_*` is the existing mechanism that preserves `device`, `dtype`, and `layout` (and supporting overwriting), and is recommended in the doc and the migration guide.

**Comment by gchanan** [comment]
@Ssnl I don't think we are arguing about the same thing: my response was in regard to adding `like` or `as` argument; sorry if that wasn't clear!  I don't have any objection to the `new_` functions except I don't think they are discoverable.

> **t.options() isn't more consistent.

I don't really agree with this, but I don't think it's important either; the argument against `as` or `like` don't depend on consistency with the C++ API.  But your argument here papers over the differences between "more consistent" and "is exactly the same as".

> **t.options() isn't clearer either.

Again, I think we've mixed the arguments up.  I don't think `**t.options()` is clearer than the `new_` functions, I think it is clearer than `like` or `as` parameters and more discoverable than the `new_` functions.

> Again, what's the use of t.options() as a public API if its use is pretty much only in kwarg expansion of tensor factory methods?

See above: " I think it is clearer than `like` or `as` parameters and more discoverable than the `new_` functions."

**Comment by SsnL** [comment]
@gchanan Ah I see. Sorry that I misunderstood your argument. Thanks for explanation!

**Comment by vadimkantorov** [comment]
One disconcerting thing about `*.new_*` is that every tensor function needs to be implemented twice for this to be truly idiomatic, just stumbled on `torch.arange`

**Pull request closed without merging**